packages/lib/server/repository/team.ts

```diff
@@ -0,0 +1,148 @@
+   */
+  forOrgWithSlug: string | null;
+  /**
+   * If true, means that we need to fetch an organization with the given slug. Otherwise, we need to fetch a team with the given slug.
+   */
+  isOrg: boolean;
+  teamSelect: TeamSelect;
+};
+
+const log = logger.getSubLogger({ prefix: ["repository", "team"] });
+
+/**
+ * Get's the team or organization with the given slug or id reliably along with parsed metadata.
+ */
+async function getTeamOrOrg<TeamSelect extends Prisma.TeamSelect>({
+  lookupBy,
+  forOrgWithSlug: forOrgWithSlug,
+  isOrg,
+  teamSelect,
+}: GetTeamOrOrgArg<TeamSelect>): Promise<TeamGetPayloadWithParsedMetadata<TeamSelect>> {
+  const where: Prisma.TeamFindFirstArgs["where"] = {};
+  teamSelect = {
+    ...teamSelect,
+    metadata: true,
+  };
+  if (lookupBy.havingMemberWithId) where.members = { some: { userId: lookupBy.havingMemberWithId } };
+
+  if ("id" in lookupBy) {
+    where.id = lookupBy.id;
+  } else {
+    where.slug = lookupBy.slug;
+  }
+
+  if (isOrg) {
+    // We must fetch only the organization here.
+    // Note that an organization and a team that doesn't belong to an organization, both have parentId null
+    // If the organization has null slug(but requestedSlug is 'test') and the team also has slug 'test', we can't distinguish them without explicitly checking the metadata.isOrganization
+    // Note that, this isn't possible now to have same requestedSlug as the slug of a team not part of an organization. This is legacy teams handling mostly. But it is still safer to be sure that you are fetching an Organization only in case of isOrgView
+    where.metadata = {
+      path: ["isOrganization"],
+      equals: true,
+    };
+    // We must fetch only the team here.
+  } else {
+    if (forOrgWithSlug) {
+      where.parent = whereClauseForOrgWithSlugOrRequestedSlug(forOrgWithSlug);
+    }
+  }
+
+  log.debug({
+    orgSlug: forOrgWithSlug,
+    teamLookupBy: lookupBy,
+    isOrgView: isOrg,
+    where,
+  });
+
+  // teamSelect extends Prisma.TeamSelect but still teams doesn't contain a valid team as per TypeScript and thus it doesn't consider it having team.metadata, team.id and other fields
+  // This is the reason below code is using a lot of assertions.
+  const teams = await prisma.team.findMany({
+    where,
+    select: teamSelect,
+  });
+
+  const teamsWithParsedMetadata = teams
+    .map((team) => ({
+      ...team,
+      // Using Type assertion here because we know that the metadata is present and Prisma and TypeScript aren't playing well together
+      metadata: teamMetadataSchema.parse((team as { metadata: z.infer<typeof teamMetadataSchema> }).metadata),
+    }))
+    // In cases where there are many teams with the same slug, we need to find out the one and only one that matches our criteria
+    .filter((team) => {
+      // We need an org if isOrgView otherwise we need a team
+      return isOrg ? team.metadata?.isOrganization : !team.metadata?.isOrganization;
```

Comment:
This is what fixes the bug. So, even if we get wrong result, we validate it here and then reject it.